src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -20,17 +20,207 @@
+    /// Merged JSONPaths type for obj1 and obj2:
+    /// obj1 ⋃ obj2 : {'a.b' : Int64, 'a.c.d' : String, 'a.f' : Array(Int64), 'e' : String, 'g.h' : Int64}
+    /// Result Named Tuple:
+    /// Tuple(a Tuple(b Int64, c Tuple(d String), f Array(Int64)), e String, g Tuple(h Int64))
+    class DataTypeJSONPaths : public IDataTypeDummy
+    {
+    public:
+        /// We create DataTypeJSONPaths on each row in input data, to
+        /// compare and merge such types faster, we use hash map to
+        /// store mapping path -> data_type. Path is a vector
+        /// of path components, to use hash map we need a hash
+        /// for std::vector<String>. We cannot just concatenate
+        /// components with '.' and store it as a string,
+        /// because components can also contain '.'
+        struct PathHash
+        {
+            size_t operator()(const std::vector<String> & path) const
+            {
+                SipHash hash;
+                hash.update(path.size());
+                for (const auto & part : path)
+                    hash.update(part);
+                return hash.get64();
+            }
+        };
+
+        using Paths = std::unordered_map<std::vector<String>, DataTypePtr, PathHash>;
+
+        explicit DataTypeJSONPaths(Paths paths_) : paths(std::move(paths_))
+        {
+        }
+
+        DataTypeJSONPaths() = default;
+
+        const char * getFamilyName() const override { return "JSONPaths"; }
+        String doGetName() const override { return finalize()->getName(); }
+        TypeIndex getTypeId() const override { return TypeIndex::JSONPaths; }
+
+        String getSQLCompatibleName() const override
+        {
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method getSQLCompatibleName is not implemented for JSONObjectForInference type");
+        }
+
+        bool isParametric() const override
+        {
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Method isParametric is not implemented for JSONObjectForInference type");
+        }
+
+        bool equals(const IDataType & rhs) const override
+        {
+            if (this == &rhs)
+                return true;
+
+            if (rhs.getTypeId() != getTypeId())
+                return false;
+
+            const auto & rhs_paths = assert_cast<const DataTypeJSONPaths &>(rhs).paths;
+            if (paths.size() != rhs_paths.size())
+                return false;
+
+            for (const auto & [path, type] : paths)
+            {
+                auto it = rhs_paths.find(path);
+                if (it == rhs_paths.end() || !it->second->equals(*type))
+                    return false;
+            }
+
+            return true;
+        }
+
+        bool merge(const DataTypeJSONPaths & rhs, std::function<void(DataTypePtr & type1, DataTypePtr & type2)> transform_types)
+        {
+            for (const auto & [rhs_path, rhs_type] : rhs.paths)
+            {
+                auto it = paths.find(rhs_path);
```

Comment:
Good idea, will try

Reply:
Done

Reply:
sorry, I wrote in my example `insert` and mentioned `try_emplace` because I found it later.
In any case, `try_emplace` seems better because it avoids copying of key and value if it's not needed.
But wait for review, if there won't be other comments we can merge as is.